components/http_client.rst

```diff
@@ -0,0 +1,530 @@
+
+When uploading JSON payloads, use the ``json`` option instead of ``body``. The
+given content will be JSON-encoded automatically and the request will add the
+``Content-Type: application/json`` automatically too::
+
+    $response = $httpClient->request('POST', 'https://...', [
+        'json' => ['param1' => 'value1', '...'],
+    ]);
+
+Cookies
+~~~~~~~
+
+The HTTP client provided by this component is stateless but handling cookies
+requires a stateful storage (because responses can update cookies and they must
+be used for subsequent requests). That's why this component doesn't handle
+cookies automatically.
+
+You can either handle cookies yourself using the ``Cookie`` HTTP header or use
+the :doc:`BrowserKit component </components/browser_kit>` which provides this
+feature and integrates seamlessly with the HttpClient component.
+
+Redirects
+~~~~~~~~~
+
+By default, the HTTP client follows redirects, up to a maximum of 20, when
+making a request. Use the ``max_redirects`` setting to configure this behavior
+(if the number of redirects is higher than the configured value, you'll get a
+:class:`Symfony\\Component\\HttpClient\\Exception\\RedirectionException`)::
+
+    $response = $httpClient->request('GET', 'https://...', [
+        // 0 means to not follow any redirect
+        'max_redirects' => 0,
+    ]);
+
+Concurrent Requests
+~~~~~~~~~~~~~~~~~~~
+
+
+.. TODO
+
+
+Asynchronous Requests
+~~~~~~~~~~~~~~~~~~~~~
+
+
+.. TODO  see https://gist.github.com/tgalopin/a84a11ece0621b8a79ed923afe015b3c
+
+
+Processing Responses
+--------------------
+
+The response returned by all HTTP clients is an object of type
+:class:`Symfony\\Contracts\\HttpClient\\ResponseInterface` which provides the
+following methods::
+
+    $response = $httpClient->request('GET', 'https://...');
+
+    // gets the HTTP status code of the response
+    $statusCode = $response->getStatusCode();
+
+    // gets the HTTP headers as string[][] with the header names lower-cased
+    $headers = $response->getHeaders();
+
+    // gets the response body as a string
+    $content = $response->getContent();
+
+    // returns info coming from the transport layer, such as "raw_headers",
+    // "redirect_count", "start_time", "redirect_url", etc.
+    $httpInfo = $response->getInfo();
+    // you can get individual info too
+    $startTime = $response->getInfo('start_time');
+
+Streaming Responses
+~~~~~~~~~~~~~~~~~~~
+
+Call to the ``stream()`` method of the HTTP client to get *chunks* of the
+response sequentially instead of waiting for the entire response::
+
+    $url = 'https://releases.ubuntu.com/18.04.1/ubuntu-18.04.1-desktop-amd64.iso';
+    $response = $httpClient->request('GET', $url, [
+        // optional: if you don't want to buffer the response in memory
+        'buffer' => false,
+        // optional: to display details about the response progress
+        'on_progress' => function (int $dlNow, int $dlSize, array $info) {
```

Comment:
Maybe we could should `void` as return type of the anonymous function for clarification.